File: Sources/CBTReframe/ViewModels/ReframeViewModel.swift
import Foundation
import SwiftUI
#if !SKIP
import SwiftData
#endif

@MainActor
@Observable
final class ReframeViewModel {
    var inputText: String = ""
    var result: AnalysisResult?
    var isLoading: Bool = false
    var errorMessage: String?
    var showCrisisBanner: Bool = false
    var isButtonPressed: Bool = false
    var retryRecoveryNotice: String?
    var streamingText: String = ""
    var isStreamingResult: Bool = false
    var latestHistoryEntryID: UUID?

    var selectedMood: String = ""
    var isAkathisia: Bool = false

    var analysisElapsedSeconds: Int = 0
    var thinkingPhraseIndex: Int = 0
    private var thinkingTickerTask: Task<Void, Never>?
    private var retryNoticeTask: Task<Void, Never>?

    var globalSettings: GlobalSettings
    private let resolver: AIProviderResolver
    private let streakService: StreakService

    var currentStreak: Int = 0
    var longestStreak: Int = 0
    var todayAnalysisCount: Int = 0

    static let quickStartPrompts: [(emoji: String, text: String)] = [
        ("💭", "我觉得自己什么都做不好"),
        ("😰", "明天开会我一定会搞砸"),
        ("😔", "没有人真正关心我"),
        ("😤", "所有事情都不顺利"),
        ("🫠", "我太累了，什么都不想做"),
    ]

    private let reframeUseCase: ReframeUseCase

    /// 分析进行中时首页加载条样式（互斥；与 `analyzeThought` 里是否启动计时器一致）。
    enum LoadingBannerStyle: Equatable {
        case none
        /// OpenAI o‑系列 / DeepSeek Reasoner 等：计时 + 阶段文案
        case deepReasoningWithTimer
        /// Gemini Pro：偏慢但无链式思考 UI，仅轻量提示
        case geminiPro
    }

    var loadingBannerStyle: LoadingBannerStyle {
        if modelIndicatesDeepReasoning { return .deepReasoningWithTimer }
        if modelIndicatesGeminiPro { return .geminiPro }
        return .none
    }

    private var modelIndicatesDeepReasoning: Bool {
        let id = resolver.selectedModel.id.lowercased()
        return id.contains("reasoner")
            || id.hasPrefix("o1") || id.hasPrefix("o3") || id.hasPrefix("o4")
            || id.contains("reason")
            || id.contains("thinking")
    }

    private var modelIndicatesGeminiPro: Bool {
        resolver.selectedProvider == .gemini
            && resolver.selectedModel.id.lowercased().contains("pro")
    }

    static let thinkingPhrases: [String] = [
        "理解中",
        "梳理中",
        "提炼中",
        "整理回复",
    ]

    var currentThinkingPhrase: String {
        Self.thinkingPhrases[thinkingPhraseIndex % Self.thinkingPhrases.count]
    }

    var suggestedThinkingTemplate: ThinkingTemplate? {
        ThinkingTemplate.suggest(for: inputText)
    }

    var activeTemplate: PromptTemplate {
        globalSettings.thinkingTemplate.promptTemplate
    }

    init(
        globalSettings: GlobalSettings,
        resolver: AIProviderResolver,
        reframeUseCase: ReframeUseCase,
        streakService: StreakService
    ) {
        self.globalSettings = globalSettings
        self.resolver = resolver
        self.reframeUseCase = reframeUseCase
        self.streakService = streakService
        loadStreak()
        todayAnalysisCount = streakService.todayAnalysisCount()
    }

    var greeting: String {
        let hour = Calendar.current.component(.hour, from: Date())
        switch hour {
        case 5..<12: return "早上好"
        case 12..<14: return "中午好"
        case 14..<18: return "下午好"
        case 18..<22: return "晚上好"
        default: return "夜深了"
        }
    }

    static let dailyQuotes: [String] = [
        "每一个想法都只是想法，不是事实。",
        "你不需要相信脑海中的每一句话。",
        "今天也在努力理解自己，这已经很了不起了。",
        "改变从觉察开始。",
        "对自己温柔一点，你正在做一件勇敢的事。",
        "情绪像天气，会变的。",
        "你可以感受到痛苦，同时选择前行。",
        "你不需要完美，只需要前进一小步。",
        "承认情绪本身就是一种力量。",
        "慢一点没关系，你已经在路上了。",
    ]

    var todayQuote: String {
        let dayOfYear = Calendar.current.ordinality(of: .day, in: .year, for: Date()) ?? 0
        return Self.dailyQuotes[dayOfYear % Self.dailyQuotes.count]
    }

    private func incrementTodayCount() {
        todayAnalysisCount = streakService.incrementTodayCount()
    }

    private func loadStreak() {
        let streak = streakService.loadStreak()
        currentStreak = streak.current
        longestStreak = streak.longest
    }

    private func markStreakToday() {
        let streak = streakService.markToday()
        currentStreak = streak.current
        longestStreak = streak.longest
    }

    /// 生成与当前设置、风险路由一致的完整提示词，供复制到外站（免 App 内 API 费用）。
    func buildExternalManualPromptText() -> String? {
        let thought = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !thought.isEmpty else { return nil }
        let mood = selectedMood.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !mood.isEmpty else { return nil }
        let strategy = routeStrategy(level: detectRiskLevel(thought))
        return PromptBuilder.buildExternalPasteboardText(
            thought: thought,
            mood: mood,
            mode: globalSettings.analysisDepth.reframeMode,
            style: globalSettings.responseStyle.legacyResponseStyle,
            template: globalSettings.thinkingTemplate.promptTemplate,
            strategy: strategy,
            hasAkathisia: isAkathisia
        )
    }

    #if !SKIP
    @MainActor
    func analyzeThought(modelContext: ModelContext) async {
        await _analyzeThought(context: modelContext)
    }
    #else
    @MainActor
    func analyzeThought() async {
        await _analyzeThought(context: nil)
    }
    #endif

    @MainActor
    private func _analyzeThought(context: Any?) async {
        let thought = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !thought.isEmpty else { return }

        let moodTrimmed = selectedMood.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !moodTrimmed.isEmpty else {
            errorMessage = "先点一个最接近现在状态的心情，再继续。"
            return
        }
        errorMessage = nil

        isLoading = true
        errorMessage = nil
        isStreamingResult = true
        streamingText = ""
        if loadingBannerStyle == LoadingBannerStyle.deepReasoningWithTimer {
            startThinkingProgress()
        }
        defer {
            stopThinkingProgress()
            isLoading = false
        }

        #if !SKIP
        if let ctx = context as? ModelContext {
            let useCaseOutput = await reframeUseCase.analyze(
                thought: thought,
                mood: moodTrimmed,
                isAkathisia: isAkathisia,
                globalSettings: globalSettings,
                modelContext: ctx
            )
            await processOutput(useCaseOutput)
        }
        #else
        let useCaseOutput = await reframeUseCase.analyze(
            thought: thought,
            mood: moodTrimmed,
            isAkathisia: isAkathisia,
            globalSettings: globalSettings,
            modelContext: ModelContext()
        )
        await processOutput(useCaseOutput)
        #endif
    }

    @MainActor
    private func processOutput(_ useCaseOutput: ReframeUseCaseOutput) async {
        showCrisisBanner = useCaseOutput.showCrisisBanner
        if let message = useCaseOutput.errorMessage {
            errorMessage = message
            return
        }
        guard let analysisResult = useCaseOutput.result else {
            errorMessage = "分析失败，请稍后重试"
            return
        }
        await playStreamingText(for: analysisResult)

        withAnimation(.spring(response: 0.5, dampingFraction: 0.8)) {
            self.result = analysisResult
        }
        isStreamingResult = false
        
        if useCaseOutput.recoveredByRetry {
            showRetryRecoveryNotice()
        }
        latestHistoryEntryID = useCaseOutput.historyEntryID

        markStreakToday()
        incrementTodayCount()
        HapticManager.success()
    }

    @MainActor
    func reset() {
        inputText = ""
        selectedMood = ""
        isAkathisia = false
        result = nil
        errorMessage = nil
        showCrisisBanner = false
        retryRecoveryNotice = nil
        retryNoticeTask?.cancel()
        retryNoticeTask = nil
        stopThinkingProgress()
        latestHistoryEntryID = nil
    }

    @MainActor
    private func startThinkingProgress() {
        analysisElapsedSeconds = 0
        thinkingPhraseIndex = 0
        thinkingTickerTask?.cancel()
        thinkingTickerTask = Task { @MainActor in
            var ticks = 0
            while !Task.isCancelled && isLoading {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                guard !Task.isCancelled, isLoading else { break }
                ticks += 1
                analysisElapsedSeconds = ticks
                if ticks % 2 == 0 {
                    thinkingPhraseIndex = (thinkingPhraseIndex + 1) % Self.thinkingPhrases.count
                }
            }
        }
    }

    @MainActor
    private func stopThinkingProgress() {
        thinkingTickerTask?.cancel()
        thinkingTickerTask = nil
        analysisElapsedSeconds = 0
        thinkingPhraseIndex = 0
    }

    @MainActor
    private func showRetryRecoveryNotice() {
        let notice = "网络波动，已自动重试并成功"
        withAnimation(.easeInOut(duration: 0.2)) {
            retryRecoveryNotice = notice
        }
        retryNoticeTask?.cancel()
        retryNoticeTask = Task { @MainActor in
            try? await Task.sleep(nanoseconds: 3_500_000_000)
            guard !Task.isCancelled else { return }
            withAnimation(.easeInOut(duration: 0.2)) {
                if retryRecoveryNotice == notice {
                    retryRecoveryNotice = nil
                }
            }
        }
    }

    @MainActor
    private func playStreamingText(for result: AnalysisResult) async {
        let full = [
            "认知扭曲：\(result.distortion)",
            "替代想法：\(result.alternative)",
            "建议行动：\(result.action)",
        ].joined(separator: "\n")
        streamingText = ""
        for ch in full {
            streamingText = streamingText + String(ch)
            try? await Task.sleep(nanoseconds: 8_000_000)
        }
    }
}

File: shared/src/commonMain/kotlin/com/henryliu/cbtreframe/shared/ReframeViewModel.kt
package com.henryliu.cbtreframe.shared

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime

data class ReframeUiState(
    val inputText: String = "",
    val selectedMood: String = "",
    val isAkathisia: Boolean = false,
    val result: AnalysisResult? = null,
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val showCrisisBanner: Boolean = false,
    val isButtonPressed: Boolean = false,
    val retryRecoveryNotice: String? = null,
    val streamingText: String = "",
    val isStreamingResult: Boolean = false,
    val latestHistoryEntryID: String? = null,
    val analysisElapsedSeconds: Int = 0,
    val thinkingPhraseIndex: Int = 0,
    val currentStreak: Int = 0,
    val longestStreak: Int = 0,
    val todayAnalysisCount: Int = 0,
    val suggestedThinkingTemplate: ThinkingTemplate? = null,
    val selectedProvider: AIProvider = AIProvider.LOCAL,
    val selectedModelName: String = "",
    val selectedTemplate: ThinkingTemplate? = null,
) {
    val greeting: String
        get() {
            val hour = kotlinx.datetime.Clock.System.now()
                .toLocalDateTime(kotlinx.datetime.TimeZone.currentSystemDefault()).hour
            return when (hour) {
                in 5..11 -> "早上好"
                in 12..13 -> "中午好"
                in 14..17 -> "下午好"
                in 18..21 -> "晚上好"
                else -> "夜深了"
            }
        }

    val todayQuote: String
        get() {
            val dayOfYear = kotlinx.datetime.Clock.System.now()
                .toLocalDateTime(kotlinx.datetime.TimeZone.currentSystemDefault())
                .dayOfYear
            return ReframeViewModel.dailyQuotes[dayOfYear % ReframeViewModel.dailyQuotes.size]
        }

    val currentThinkingPhrase: String
        get() = ReframeViewModel.thinkingPhrases[thinkingPhraseIndex % ReframeViewModel.thinkingPhrases.size]

    val isDeepReasoningModel: Boolean
        get() {
            val id = selectedModelName.lowercase()
            return id.contains("reasoner")
                || id.startsWith("o1") || id.startsWith("o3") || id.startsWith("o4")
                || id.contains("reason")
                || id.contains("thinking")
        }

    val isGeminiProModel: Boolean
        get() {
            val id = selectedModelName.lowercase()
            return id.contains("gemini") && id.contains("pro")
        }

    val loadingBannerStyle: LoadingBannerStyle
        get() = when {
            isDeepReasoningModel -> LoadingBannerStyle.DEEP_REASONING
            isGeminiProModel -> LoadingBannerStyle.GEMINI_PRO
            else -> LoadingBannerStyle.NONE
        }

    val homeStage: HomeStage
        get() = when {
            inputText.isBlank() -> HomeStage.QuickStart
            selectedTemplate == null -> HomeStage.WritingThought
            selectedMood.isBlank() -> HomeStage.ChoosingMood
            else -> HomeStage.ReviewReady
        }
}

enum class LoadingBannerStyle { NONE, DEEP_REASONING, GEMINI_PRO }

class ReframeViewModel(
    private val useCase: ReframeUseCase,
    private val streakService: StreakService,
    private val settingsManager: SettingsManager,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main),
) {
    private val _uiState = MutableStateFlow(ReframeUiState())
    val uiState: StateFlow<ReframeUiState> = _uiState.asStateFlow()

    private var thinkingTickerJob: Job? = null
    private var retryNoticeJob: Job? = null

    companion object {
        val quickStartPrompts: List<Pair<String, String>> = listOf(
            "💭" to "我觉得自己什么都做不好",
            "😰" to "明天开会我一定会搞砸",
            "😔" to "没有人真正关心我",
            "😤" to "所有事情都不顺利",
            "🫠" to "我太累了，什么都不想做",
        )

        val thinkingPhrases: List<String> = listOf(
            "理解中",
            "梳理中",
            "提炼中",
            "整理回复",
        )

        val dailyQuotes: List<String> = listOf(
            "每一个想法都只是想法，不是事实。",
            "你不需要相信脑海中的每一句话。",
            "今天也在努力理解自己，这已经很了不起了。",
            "改变从觉察开始。",
            "对自己温柔一点，你正在做一件勇敢的事。",
            "情绪像天气，会变的。",
            "你可以感受到痛苦，同时选择前行。",
            "你不需要完美，只需要前进一小步。",
            "承认情绪本身就是一种力量。",
            "慢一点没关系，你已经在路上了。",
        )
    }

    init {
        val provider = settingsManager.getSelectedProvider()
        val modelName = settingsManager.getSelectedModelId().ifEmpty {
            provider.defaultModelId()
        }
        _uiState.value = _uiState.value.copy(
            selectedProvider = provider,
            selectedModelName = modelName,
        )
        loadStreak()
        val count = streakService.todayAnalysisCount()
        _uiState.value = _uiState.value.copy(todayAnalysisCount = count)
    }

    fun clear() {
        scope.cancel()
    }

    // ── Input state mutations ──────────────────────────────────────────

    fun setInputText(text: String) {
        val template = ThinkingTemplate.suggest(text)
        _uiState.value = _uiState.value.copy(
            inputText = text,
            suggestedThinkingTemplate = template,
        )
    }

    fun setSelectedMood(mood: String) {
        _uiState.value = _uiState.value.copy(selectedMood = mood)
    }

    fun setAkathisia(value: Boolean) {
        _uiState.value = _uiState.value.copy(isAkathisia = value)
    }

    fun setSelectedTemplate(template: ThinkingTemplate) {
        _uiState.value = _uiState.value.copy(selectedTemplate = template)
    }

    fun refreshProviderAndModel() {
        val provider = settingsManager.getSelectedProvider()
        val modelName = settingsManager.getSelectedModelId().ifEmpty {
            provider.defaultModelId()
        }
        _uiState.value = _uiState.value.copy(
            selectedProvider = provider,
            selectedModelName = modelName,
        )
    }

    // ── Analyze ────────────────────────────────────────────────────────

    fun analyzeThought(globalSettings: GlobalSettings = GlobalSettings.Default) {
        val thought = _uiState.value.inputText.trim()
        if (thought.isEmpty()) return

        val mood = _uiState.value.selectedMood.trim()
        if (mood.isEmpty()) {
            _uiState.value = _uiState.value.copy(
                errorMessage = "先点一个最接近现在状态的心情，再继续。"
            )
            return
        }

        val isAkathisia = _uiState.value.isAkathisia
        val provider = _uiState.value.selectedProvider
        val modelName = _uiState.value.selectedModelName

        _uiState.value = _uiState.value.copy(
            isLoading = true,
            errorMessage = null,
            isStreamingResult = true,
            streamingText = "",
            showCrisisBanner = false,
        )

        val isDeepReasoning = _uiState.value.isDeepReasoningModel
        if (isDeepReasoning) startThinkingProgress()

        scope.launch {
            try {
                val output = useCase.streamAnalyze(
                    thought = thought,
                    mood = mood,
                    hasAkathisia = isAkathisia,
                    provider = provider,
                    modelName = modelName,
                    settings = globalSettings,
                )

                if (output.showCrisisBanner) {
                    _uiState.value = _uiState.value.copy(showCrisisBanner = true)
                }

                var currentText = ""
                output.stream.collect { chunk ->
                    currentText += chunk
                    
                    // Simple hack to hide JSON structure from the raw stream 
                    // until we properly implement markdown-streaming prompt
                    val cleaned = cleanStreamContent(currentText)
                    _uiState.value = _uiState.value.copy(streamingText = cleaned)
                }

                // Wait, parseReframeOutput is already used inside the use case 
                // so we can just parse it here again or we could have returned it 
                // from the use case. But we don't have the final parsed result easily
                // returned from streamAnalyze unless we change the signature again.
                // For MVP: parse the full collected JSON here again to display the final result.
                val finalParsed = try {
                    parseReframeOutput(currentText, output.strategy).normalized(globalSettings.thinkingTemplate)
                } catch(e: Exception) {
                    AnalysisResult(
                        distortion = "分析结束",
                        alternative = _uiState.value.streamingText,
                        action = ""
                    )
                }

                _uiState.value = _uiState.value.copy(
                    result = finalParsed,
                    isStreamingResult = false,
                    latestHistoryEntryID = output.historyEntryID,
                )

                // No need to check recoveredByRetry since streaming bypasses Validation client
                
                markStreakToday()
                incrementTodayCount()

            } catch (e: CancellationException) {
                // ignore
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    errorMessage = e.message ?: "分析失败，请稍后重试",
                )
            } finally {
                stopThinkingProgress()
                _uiState.value = _uiState.value.copy(isLoading = false)
            }
        }
    }

    fun reset() {
        thinkingTickerJob?.cancel()
        thinkingTickerJob = null
        retryNoticeJob?.cancel()
        retryNoticeJob = null
        stopThinkingProgress()

        _uiState.value = _uiState.value.copy(
            inputText = "",
            selectedMood = "",
            isAkathisia = false,
            result = null,
            errorMessage = null,
            showCrisisBanner = false,
            retryRecoveryNotice = null,
            isStreamingResult = false,
            streamingText = "",
            latestHistoryEntryID = null,
        )
    }

    // ── External prompt (for copy-to-clipboard) ────────────────────────

    fun buildExternalManualPromptText(): String? {
        val thought = _uiState.value.inputText.trim()
        if (thought.isEmpty()) return null
        val mood = _uiState.value.selectedMood.trim()
        if (mood.isEmpty()) return null

        val riskLevel = detectRiskLevel(thought)
        val strategy = routeStrategy(riskLevel)

        return PromptBuilder.buildSystemPrompt(
            template = GlobalSettings.Default.thinkingTemplate,
            strategy = strategy,
            depth = GlobalSettings.Default.analysisDepth,
            style = GlobalSettings.Default.responseStyle,
            mood = mood,
            hasAkathisia = _uiState.value.isAkathisia,
        ) + "\n\n" + PromptBuilder.buildUserPrompt(thought, mood, _uiState.value.isAkathisia)
    }

    // ── Streak helpers ─────────────────────────────────────────────────

    private fun loadStreak() {
        val streak = streakService.loadStreak()
        _uiState.value = _uiState.value.copy(
            currentStreak = streak.current,
            longestStreak = streak.longest,
        )
    }

    private fun markStreakToday() {
        val streak = streakService.markToday()
        _uiState.value = _uiState.value.copy(
            currentStreak = streak.current,
            longestStreak = streak.longest,
        )
    }

    private fun incrementTodayCount() {
        val count = streakService.incrementTodayCount()
        _uiState.value = _uiState.value.copy(todayAnalysisCount = count)
    }

    // ── Thinking progress ticker ───────────────────────────────────────

    private fun startThinkingProgress() {
        _uiState.value = _uiState.value.copy(analysisElapsedSeconds = 0, thinkingPhraseIndex = 0)
        thinkingTickerJob?.cancel()
        thinkingTickerJob = scope.launch {
            var ticks = 0
            while (isActive && _uiState.value.isLoading) {
                delay(1000)
                if (!isActive || !_uiState.value.isLoading) break
                ticks++
                _uiState.value = _uiState.value.copy(
                    analysisElapsedSeconds = ticks,
                    thinkingPhraseIndex = if (ticks % 2 == 0)
                        (_uiState.value.thinkingPhraseIndex + 1) % thinkingPhrases.size
                    else _uiState.value.thinkingPhraseIndex,
                )
            }
        }
    }

    private fun stopThinkingProgress() {
        thinkingTickerJob?.cancel()
        thinkingTickerJob = null
        _uiState.value = _uiState.value.copy(analysisElapsedSeconds = 0, thinkingPhraseIndex = 0)
    }

    // ── Retry recovery notice ──────────────────────────────────────────

    private fun showRetryRecoveryNotice() {
        val notice = "网络波动，已自动重试并成功"
        _uiState.value = _uiState.value.copy(retryRecoveryNotice = notice)
        retryNoticeJob?.cancel()
        retryNoticeJob = scope.launch {
            delay(3500)
            if (isActive && _uiState.value.retryRecoveryNotice == notice) {
                _uiState.value = _uiState.value.copy(retryRecoveryNotice = null)
            }
        }
    }
}

// ── ThinkingTemplate.suggest extension ─────────────────────────────────────

fun ThinkingTemplate.Companion.suggest(text: String): ThinkingTemplate? {
    val lower = text.trim().lowercase()
    return when {
        lower.length <= 5 -> null
        lower.contains("为什么") || lower.contains("怎么会") || lower.contains("怎么办")
            -> ThinkingTemplate.socratic
        lower.contains("做") || lower.contains("去") || lower.contains("行动") || lower.contains("试试")
            -> ThinkingTemplate.behavioral
        lower.contains("觉得") || lower.contains("认为") || lower.contains("感觉") || lower.contains("想")
            -> ThinkingTemplate.cbt
        else -> null
    }
}

// ── Stream Content Cleaner ──────────────────────────────────────────────────

fun cleanStreamContent(currentText: String): String {
    return currentText
        .replace(Regex("\"?[a-zA-Z_]+\"\\s*:\\s*\"?"), "")
        .replace(Regex("[{}\\[\\]]"), "")
        .replace("\\n", "\n")
        .replace("\\\"", "\"")
        .trim()
}
